### ISSUE TITLE
Easier cactbot customization

### ISSUE THREAD

**Issue by quisquous** [quisquous/cactbot#155]
From the discussion on #152, I wanted to make a feature here to discuss making cactbot customization easier.  It'd be helpful to know what kinds of customization folks find themselves doing a lot.  I'd also love to know what's hard (and you wish was easier), and what's impossible that you wish could be possible.

**Comment by zagibara** [comment]
I wish i could move timeline and popup text separately on my screen instead of editing .css file

**Comment by EricSpeidel** [comment]
Not sure if that's limited by the OverlayPlugin, but resizing (Dragging corners to resize) for OopsyRaidsy would be sweet without doing css. IIRC the Kagerou parser skin does this.

**Comment by mooondark** [comment]
I wish I could resize the timebars because sometimes, the translation does not fit.

**Comment by chocoreb** [comment]
I don't know if this is possible within Cactbot, but it would be really awesome if I could choose an audio device to output its trigger sounds to! (like usb headphones or voice meeter)  The SlimDX Audio plugin doesn't seem to do the job reliably :(

**Comment by quisquous** [comment]
@chocoreb Take a look at #120.  CheVolume might do what you want.

**Comment by mooondark** [comment]
It would be nice to add an option directly in the GUI to change X,Y coordinates. I play on 2 screens and sometimes I switch to one screen only.
If, for exemple, the raidboss window is on the other screen, I can't move it easilly.

**Comment by meciuPL** [comment]
Changing the size, colours of the bars and the text inside as well as the font text and its formatting. Basic stuff. Separating timeline from popup would be nice too.

**Comment by ArkfrostLumas** [comment]
As a contributor myself the only request I have is a button or option within ACT (within the cactbot overlay plugin) to switch languages or that cactbot automatically switches to the language set in the FFXIV-Parsing-Plugin (which everyone needs anyway for ACT to work in general) :)

**Comment by quisquous** [comment]
@meciuPL You should be able to do all of that now with CSS customization.  Are you asking for that to be easier?

**Comment by Wyti** [comment]
I wish I could easily choose which direction to grow when new bars are added (i.e fix the popup position and have the bars grow upward when needed), as well as inverting the sorting order (independently of the grow direction).

**Comment by TKDBB84** [comment]
It would be really nice to be able to toggle different sounds on an off, as an example the "game on" for ready check I previously would always disable by removing in the JSON array.  Now that the files are hosted remotely by default; it takes more effort to remove.

**Comment by KMontag42** [comment]
i think a lot of people would benefit from a guided styling experience instead of having to overwrite the css files.

[kagerou's](https://github.com/hibiyasleep/kagerou) overlay does a great job of walking non technical people through styling the UI, and also has an `advanced` tab which lets power users update the css itself directly.

**Comment by quisquous** [comment]
Individual triggers can be toggled off.  You just need to add some lines into your local **cactbot/user/raidboss.js** file, even when you're using remote files.  See https://github.com/quisquous/cactbot/blob/master/AdvancedCactbot.md#customizing-behavior for more details.

In particular, add these lines:
```
Options.DisabledTriggers = {
  'General Ready check': true,
}
```

**Comment by TKDBB84** [comment]
That's cool, i'll do that.  Thank you @quisquous .  Along the lines of the issue title "Easier Cactbot customization"; an interface tab on ACT to do that would be super cool; however I also understand that is a non-trivial amount of effort.

Again, thank you for the link for customization on the remote files 👍

**Comment by unfaiyted** [comment]
I was also thinking about the customization functionality and starting writing a small UI today for the primary settings currently I broke it out from the primary addons and created another module to support this. Here is the start of what I have so far: 
https://gyazo.com/ccb2875bfb6d9379aeff2c824aaa1831

I also have a branch I am trying to work this out of myself, also wondering about feedback and stuff or if something like this is even useful for people

**Comment by quisquous** [comment]
@unfaiyted :heart_eyes: Oh wow, that's some nice-looking UI.  I would be super supportive of this kind of feature.  If you want to file a separate issue that we can talk about this config UI there without crowding out this more brainstormy issue, that might be a good way to keep it focused and move forward.

I think the real catch is the question of how you save all of the settings.  Maybe we would need to send some json back to C# and write that out on disk.  Maybe something that writes out **cactbot/user/raidboss-autoconfig.js** so that folks can still do manual customization, but this UI could generate the same sort of input that folks are now doing manually.  That's just a thought, though.  Maybe you have some alternative idea of how to persist and reload this data.

**Comment by Serin-leetee** [comment]
Based on some comments i've read so far, is there a way to change the sound for something like Ready Check using the **cactbot/user/** folder? I noticed there's an option to disable the trigger, but not one to change it. I've made it work by changing the **cactbot/ui/raidboss/data/triggers/general.js** file to get it to play a custom sound, but when I tried putting a copy of general.js in my **cactbot/user/**, it doesn't work.

Is there a current implementation for this? I could just add the following lines into **cactbot/ui/raidboss/raidboss.js**
`ReadyCheckSound: '../../resources/sounds/<name>/<name.ogg>'`
and maybe get it to work that way but that's pretty much the same as changing general.js.

**Comment by quisquous** [comment]
@Serin-leetee Yes, you need to do something like https://github.com/quisquous/cactbot/blob/master/user/raidboss-example.js#L198.  Let me know if you run into any issues.

**Comment by Serin-leetee** [comment]
Its not quite working, here's what I'm doing:
I believe the id for Ready check is 'General Ready Check' so in the user folder i'm using this to test it. also a side, the Test Laugh trigger that this was based on also doesn't work for me. The test poke tts works though.
```
  'General Ready Check': {
    SpeechAlert: false,
    SoundAlert: true,
    TextAlert: false,
    SoundOverride: '../../resources/sounds/WeakAuras/CowMooing.ogg',
    VolumeOverride: 0.2,
  },
```

**Comment by Serin-leetee** [comment]
Oh that did the trick. Dang I should have noticed it earlier but I totally didn't recognize the lowercase c in the general.js file.

**Comment by altissia86** [comment]
i like to customize some of the callouts like my group in a12s doesnt use triangle and box we just move to our spot or in a11s i like to chase the head and do left right from there. i dont know how to edit the .js, right now i just disable it and do a tts from act itself

the other thing is i want to change the style of timelines but i dont know how to edit.

the best i can mention is kagerou overlay. it's noob friendly. they have coresponding tabs for it.

**Comment by Zinterax** [comment]
I'd like it if I could put my own trigger files in the user folder to override ones in the data/triggers folder. I haven't played in a while so I don't know if this feature already exists or not.

Typically when a new boss comes out I would write my own trigger files and once I'm used to them, I don't really want to relearn new callouts from the ones pushed on github. 

Typically I just back mine up and after updating to the latest release I just then overwrite all the trigger files with my own, but it would be nice to just be able to extract over everything and not have to do it myself every time.

**Comment by quisquous** [comment]
@Zinterax I think the easiest thing to do to support that would be for you to add some lines to your user raidboss file to disable all the normal triggers, e.g. https://github.com/quisquous/cactbot/blob/master/AdvancedCactbot.md#customizing-behavior.  Then you can copy and paste the whole trigger block that you would have put into **ui/raidboss/data/triggers/something.js** into the `Options.Triggers` array, like https://github.com/quisquous/cactbot/blob/master/AdvancedCactbot.md#adding-custom-triggers.

Maybe in the future with @unfaiyted's config generator, it will be easier to disable all the triggers at once.

**Comment by veltall** [comment]
As a user with customized triggers, I would like to have a streamlined process for updating Cactbot and maintaining my previous customization.

As a technical user, I can see 3 options for this:
#####Option A
1. Edit directly on the "source" data files (under `cactbot/ui/raidboss/data/[patch]/[raidtype]/file.js`),
2. Commit these intentional changes to git, e.g. branch 'custom'
3. Download cactbot and overwrite the old `cactbot` folder with the new one.
4. Branch, commit new cactbot changes, e.g. branch 'new'
5. Merge 'custom' onto 'new'

#####Option B
1. Edit every custom changes in the `~/user/raidboss.js` config file. This creates for a huge file that's impossible to read over time
2. Lay down
3. Try not to cry
4. Cry a lot

#####Option C
1. Edit each encounter's customized changes in its own `~/user/encounter.js`
2. `import` these into the main `~/user/raidboss.js` config file

It'd be nice if we could have some feedback, suggestions, and guidance on this issue (especially for newer users).

**Comment by veltall** [comment]
Personally I find it easier as a "new user" to be modifying existing words as opposed to writing on a blank sheet. Option C is clearly the cleanest and most desirable approach, but there's that moment when you open the new file and just aren't really sure what to write.

I think a hybrid approach for A and C would be best. Either something like copying the original `.js` trigger file to the `~/user` folder then start hacking away at it, or copy-pasting the relevant sections into a pre-made template then modify as desired.

**Comment by quisquous** [comment]
Out of curiosity, what is most of your customization? New triggers? Disabling triggers? Changing text on triggers?

One thing that might help is that I landed a new config UI which will be in the next release: https://github.com/quisquous/cactbot/pull/880#issuecomment-568863244

> ### Option A
> 1. Edit directly on the "source" data files (under `cactbot/ui/raidboss/data/[patch]/[raidtype]/file.js`),
> 2. Commit these intentional changes to git, e.g. branch 'custom'
> 3. Download cactbot and overwrite the old `cactbot` folder with the new one.
> 4. Branch, commit new cactbot changes, e.g. branch 'new'
> 5. Merge 'custom' onto 'new'

If you are doing this, you should add quisquous/cactbot as a git remote and steps 345 should just be to `git pull` and merge into your current branch.

I don't know that I'd recommend doing this approach, but if you're comfortable with resolving git conflicts, it's certainly an option.

> ### Option B
> 1. Edit every custom changes in the `~/user/raidboss.js` config file. This creates for a huge file that's impossible to read over time
> 2. Lay down
> 3. Try not to cry
> 4. Cry a lot

Sorry that this is such a pain.

Regarding the huge file, I think the right fix for that is something like https://github.com/quisquous/cactbot/issues/860 where the user folder could support subfolders.  This is something I would really like to have done, but haven't gotten to it yet.  Please feel free to contribute if you have time and are interested.  (I think this would address Option C for you as well.)

Would that solve your pain, or do you need more than that?

**Comment by veltall** [comment]
Most of the mods involve overriding the tts text and changing the trigger conditions so that it's more useful (and less confusing) for a group setting. Cactbot trigger files are written for personal consumption, so a lot of it needs to be changed, and GroupTTS is just one of those things. I can't override condition in the config file, so my personal use case requires direct modification of the source files.

Example of a big mod: Limit Cut in Cactbot reads the number for the user and gives them a 5s warning before they get cleaved. A group would then benefit from a 5s warning before **every pair of cleaves**.

I'm personally fine with managing versions of the files, git conflict isn't a big problem since I'll just opt to keep my version over any updates until things break. I just feel that the use case of Cactbot as a group alert is only going to get more popular and other people won't be able to cope.

> Regarding the huge file, I think the right fix for that is something like #860 where the user folder could support subfolders.

This is what I have in mind for Option C as well. I haven't looked into the source code for cactbot too deeply since I lack proficiency in js, but I can certainly try to cook something up once I understand how the resources are loaded.

**Comment by quisquous** [comment]
I think this is fixed now with the config panel.
